File: server/app/routes/translation.py
"""
Translation queue API routes.

Endpoints:
- POST /api/translations - Submit new translation
- GET /api/translations - List translations (for Translations page)
- GET /api/translations/{translation_id} - Get translation details and results
- GET /api/translations/{translation_id}/status - Quick status check
- DELETE /api/translations/{translation_id} - Delete a translation
- GET /api/translations/{translation_id}/stream - SSE stream for translation progress
"""

import asyncio
import json
from typing import Any

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from app.models import (
    CreateTranslationRequest,
    CreateTranslationResponse,
    OkResponse,
    ParagraphResult,
    TranslationDetailResponse,
    TranslationStatusResponse,
    TranslationSummary,
    ListTranslationsResponse,
    TranslationResult,
)
from app.persistence import (
    delete_translation,
    get_translation,
    get_translation_segment_count,
    get_translation_with_results,
    list_translations,
)
from app.queue import get_queue_manager

router = APIRouter(tags=["translations"])


def _translation_to_summary(translation: Any) -> TranslationSummary:
    """Convert a TranslationRecord to TranslationSummary."""
    # Get segment counts
    completed, total = (
        get_translation_segment_count(translation.id) if translation.status != "pending" else (None, None)
    )

    return TranslationSummary(
        id=translation.id,
        created_at=translation.created_at,
        status=translation.status,
        source_type=translation.source_type,
        input_preview=translation.input_text[:100] + "..."
        if len(translation.input_text) > 100
        else translation.input_text,
        full_translation_preview=(
            translation.full_translation[:100] + "..."
            if translation.full_translation and len(translation.full_translation) > 100
            else translation.full_translation
        ),
        segment_count=completed,
        total_segments=total,
    )


# --- JSON API Endpoints (prefix: /api) ---


@router.post("/api/translations", response_model=CreateTranslationResponse)
async def api_create_translation(request: CreateTranslationRequest):
    """
    Submit a new translation.

    The translation is created immediately with 'pending' status.
    Use GET /api/translations/{translation_id}/status to check progress.
    Use GET /api/translations/{translation_id}/stream to stream progress via SSE.
    """
    if not request.input_text.strip():
        raise HTTPException(status_code=400, detail="input_text is required")

    manager = get_queue_manager()
    translation_id = manager.submit_translation(
        input_text=request.input_text,
        source_type=request.source_type,
    )

    return CreateTranslationResponse(translation_id=translation_id, status="pending")


@router.get("/api/translations", response_model=ListTranslationsResponse)
async def api_list_translations(
    limit: int = 20,
    offset: int = 0,
    status: str | None = None,
):
    """
    List translations for the Translations page.

    Supports pagination and optional status filtering.
    """
    if status and status not in {"pending", "processing", "completed", "failed"}:
        raise HTTPException(status_code=400, detail="Invalid status filter")

    translations, total = list_translations(limit=limit, offset=offset, status=status)

    return ListTranslationsResponse(
        translations=[_translation_to_summary(t) for t in translations],
        total=total,
    )


@router.get("/api/translations/{translation_id}", response_model=TranslationDetailResponse)
async def api_get_translation(translation_id: str):
    """
    Get translation with full results.

    Returns the translation details and all translated segments organized by paragraph.
    """
    result = get_translation_with_results(translation_id)
    if result is None:
        raise HTTPException(status_code=404, detail="Translation not found")

    # Convert to response format
    paragraphs = None
    if result.paragraphs:
        paragraphs = [
            ParagraphResult(
                translations=[
                    TranslationResult(
                        segment=t["segment"],
                        pinyin=t["pinyin"],
                        english=t["english"],
                    )
                    for t in p["translations"]
                ],
                indent=p["indent"],
                separator=p["separator"],
            )
            for p in result.paragraphs
        ]

    return TranslationDetailResponse(
        id=result.translation.id,
        created_at=result.translation.created_at,
        status=result.translation.status,
        source_type=result.translation.source_type,
        input_text=result.translation.input_text,
        full_translation=result.translation.full_translation,
        error_message=result.translation.error_message,
        paragraphs=paragraphs,
    )


@router.get("/api/translations/{translation_id}/status", response_model=TranslationStatusResponse)
async def api_get_translation_status(translation_id: str):
    """
    Quick status check for a translation.

    Returns current status and progress without full results.
    """
    translation = get_translation(translation_id)
    if translation is None:
        raise HTTPException(status_code=404, detail="Translation not found")

    progress, total = (
        get_translation_segment_count(translation_id) if translation.status != "pending" else (None, None)
    )

    return TranslationStatusResponse(
        translation_id=translation_id,
        status=translation.status,
        progress=progress,
        total=total,
    )


@router.delete("/api/translations/{translation_id}", response_model=OkResponse)
async def api_delete_translation(translation_id: str):
    """Delete a translation and its results."""
    if not delete_translation(translation_id):
        raise HTTPException(status_code=404, detail="Translation not found")

    return OkResponse()


# --- SSE Streaming Endpoint ---


@router.get("/api/translations/{translation_id}/stream")
async def translation_stream(translation_id: str):
    """
    SSE stream for translation progress.

    Events:
    - start: { type: "start", translation_id, total, paragraphs }
    - progress: { type: "progress", current, total, result }
    - complete: { type: "complete", paragraphs, full_translation }
    - error: { type: "error", message }

    This endpoint starts processing the translation if it's pending.
    """

    async def generate():
        translation = get_translation(translation_id)
        if translation is None:
            yield f"data: {json.dumps({'type': 'error', 'message': 'Translation not found'})}\n\n"
            return

        manager = get_queue_manager()

        # If translation is already completed, send results immediately
        if translation.status == "completed":
            result = get_translation_with_results(translation_id)
            if result:
                # Calculate total segments
                total_segments = (
                    sum(len(p["translations"]) for p in result.paragraphs)
                    if result.paragraphs
                    else 0
                )

                # Send start event
                paragraph_info = (
                    [
                        {
                            "segment_count": len(p["translations"]),
                            "indent": p["indent"],
                            "separator": p["separator"],
                        }
                        for p in result.paragraphs
                    ]
                    if result.paragraphs
                    else []
                )

                yield f"data: {json.dumps({'type': 'start', 'translation_id': translation_id, 'total': total_segments, 'paragraphs': paragraph_info, 'fullTranslation': result.translation.full_translation})}\n\n"

                # Send all progress events at once
                global_idx = 0
                for para_idx, para in enumerate(result.paragraphs or []):
                    for seg_idx, t in enumerate(para["translations"]):
                        result_data = {
                            "segment": t["segment"],
                            "pinyin": t["pinyin"],
                            "english": t["english"],
                            "index": global_idx,
                            "paragraph_index": para_idx,
                        }
                        global_idx += 1
                        yield f"data: {json.dumps({'type': 'progress', 'current': global_idx, 'total': total_segments, 'result': result_data})}\n\n"

                # Send complete event
                yield f"data: {json.dumps({'type': 'complete', 'paragraphs': result.paragraphs, 'fullTranslation': result.translation.full_translation})}\n\n"
            return

        # If translation failed, send error
        if translation.status == "failed":
            yield f"data: {json.dumps({'type': 'error', 'message': translation.error_message or 'Translation failed'})}\n\n"
            return

        # Start processing if pending
        if translation.status == "pending":
            # Use a queue to collect progress updates
            progress_queue: asyncio.Queue = asyncio.Queue()

            def progress_callback(tid: str, seg_result):
                # Put progress update in queue
                try:
                    loop = asyncio.get_event_loop()
                    loop.call_soon_threadsafe(
                        progress_queue.put_nowait,
                        {
                            "type": "progress",
                            "translation_id": tid,
                            "result": seg_result,
                        },
                    )
                except Exception:
                    pass

            # Start processing in background
            manager.start_processing(translation_id, progress_callback)

            # Wait for processing to initialize
            await asyncio.sleep(0.5)

        # Poll for progress updates
        last_progress = 0
        sent_start = False

        while True:
            # Get current progress from manager
            progress = manager.get_progress(translation_id)
            if progress is None:
                # Translation might be done, check DB
                translation = get_translation(translation_id)
                if translation is None:
                    yield f"data: {json.dumps({'type': 'error', 'message': 'Translation not found'})}\n\n"
                    return
                if translation.status == "completed":
                    break
                if translation.status == "failed":
                    yield f"data: {json.dumps({'type': 'error', 'message': translation.error_message or 'Translation failed'})}\n\n"
                    return
                await asyncio.sleep(0.2)
                continue

            # Send start event once we have total
            if not sent_start and progress.get("total", 0) > 0:
                total = progress["total"]
                # Build paragraph info from results so far
                yield f"data: {json.dumps({'type': 'start', 'translation_id': translation_id, 'total': total, 'paragraphs': []})}\n\n"
                sent_start = True

            # Send progress events for new results
            current = progress.get("current", 0)
            results = progress.get("results", [])

            for i in range(last_progress, current):
                if i < len(results):
                    seg = results[i]
                    result_data = {
                        "segment": seg.segment,
                        "pinyin": seg.pinyin,
                        "english": seg.english,
                        "index": seg.global_idx,
                        "paragraph_index": seg.paragraph_idx,
                    }
                    yield f"data: {json.dumps({'type': 'progress', 'current': i + 1, 'total': progress.get('total', 0), 'result': result_data})}\n\n"

            last_progress = current

            # Check if completed
            if progress.get("status") == "completed":
                break

            if progress.get("status") == "failed":
                yield f"data: {json.dumps({'type': 'error', 'message': progress.get('error', 'Translation failed')})}\n\n"
                return

            await asyncio.sleep(0.1)

        # Send complete event
        result = get_translation_with_results(translation_id)
        if result:
            yield f"data: {json.dumps({'type': 'complete', 'paragraphs': result.paragraphs, 'fullTranslation': result.translation.full_translation})}\n\n"

        # Cleanup progress tracking
        manager.cleanup_progress(translation_id)

    return StreamingResponse(generate(), media_type="text/event-stream")

File: server/internal/http/handlers/translation.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/anath2/language-app/internal/translation"
)

type createTranslationRequest struct {
	InputText  string `json:"input_text"`
	SourceType string `json:"source_type"`
}

type createTranslationResponse struct {
	TranslationID string `json:"translation_id"`
	Status        string `json:"status"`
}

type translationSummary struct {
	ID                     string  `json:"id"`
	CreatedAt              string  `json:"created_at"`
	Status                 string  `json:"status"`
	SourceType             string  `json:"source_type"`
	InputPreview           string  `json:"input_preview"`
	FullTranslationPreview *string `json:"full_translation_preview"`
	SegmentCount           *int    `json:"segment_count"`
	TotalSegments          *int    `json:"total_segments"`
}

type listTranslationsResponse struct {
	Translations []translationSummary `json:"translations"`
	Total        int                  `json:"total"`
}

type translationDetailResponse struct {
	ID              string      `json:"id"`
	CreatedAt       string      `json:"created_at"`
	Status          string      `json:"status"`
	SourceType      string      `json:"source_type"`
	InputText       string      `json:"input_text"`
	FullTranslation *string     `json:"full_translation"`
	ErrorMessage    *string     `json:"error_message"`
	Paragraphs      interface{} `json:"paragraphs"`
}

type translationStatusResponse struct {
	TranslationID string `json:"translation_id"`
	Status        string `json:"status"`
	Progress      *int   `json:"progress"`
	Total         *int   `json:"total"`
}

func CreateTranslation(w http.ResponseWriter, r *http.Request) {
	if err := validateDependencies(); err != nil {
		WriteJSON(w, http.StatusInternalServerError, map[string]string{"detail": err.Error()})
		return
	}

	var req createTranslationRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"detail": "Invalid JSON payload"})
		return
	}

	item, err := sharedStore.Create(req.InputText, req.SourceType)
	if err != nil {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"detail": err.Error()})
		return
	}

	WriteJSON(w, http.StatusOK, createTranslationResponse{
		TranslationID: item.ID,
		Status:        item.Status,
	})

	sharedQueue.Submit(item.ID)
	sharedQueue.StartProcessing(item.ID)
}

func ListTranslations(w http.ResponseWriter, r *http.Request) {
	if err := validateDependencies(); err != nil {
		WriteJSON(w, http.StatusInternalServerError, map[string]string{"detail": err.Error()})
		return
	}

	query := r.URL.Query()
	limit := parseIntDefault(query.Get("limit"), 20)
	offset := parseIntDefault(query.Get("offset"), 0)
	status := strings.TrimSpace(query.Get("status"))

	items, total, err := sharedStore.List(limit, offset, status)
	if err != nil {
		WriteJSON(w, http.StatusBadRequest, map[string]string{"detail": err.Error()})
		return
	}

	summaries := make([]translationSummary, 0, len(items))
	for _, item := range items {
		summaries = append(summaries, translationSummary{
			ID:                     item.ID,
			CreatedAt:              item.CreatedAt,
			Status:                 item.Status,
			SourceType:             item.SourceType,
			InputPreview:           preview(item.InputText, 100),
			FullTranslationPreview: previewPtr(item.FullTranslation, 100),
			SegmentCount:           intPtrIfKnown(item.Progress, item.Status),
			TotalSegments:          intPtrIfKnown(item.Total, item.Status),
		})
	}

	WriteJSON(w, http.StatusOK, listTranslationsResponse{
		Translations: summaries,
		Total:        total,
	})
}

func GetTranslation(w http.ResponseWriter, r *http.Request) {
	if err := validateDependencies(); err != nil {
		WriteJSON(w, http.StatusInternalServerError, map[string]string{"detail": err.Error()})
		return
	}

	translationID := pathParam(r, "translation_id")
	item, ok := sharedStore.Get(translationID)
	if !ok {
		WriteJSON(w, http.StatusNotFound, map[string]string{"detail": "Translation not found"})
		return
	}

	WriteJSON(w, http.StatusOK, translationDetailResponse{
		ID:              item.ID,
		CreatedAt:       item.CreatedAt,
		Status:          item.Status,
		SourceType:      item.SourceType,
		InputText:       item.InputText,
		FullTranslation: item.FullTranslation,
		ErrorMessage:    item.ErrorMessage,
		Paragraphs:      item.Paragraphs,
	})
}

func GetTranslationStatus(w http.ResponseWriter, r *http.Request) {
	if err := validateDependencies(); err != nil {
		WriteJSON(w, http.StatusInternalServerError, map[string]string{"detail": err.Error()})
		return
	}

	translationID := pathParam(r, "translation_id")
	item, ok := sharedStore.Get(translationID)
	if !ok {
		WriteJSON(w, http.StatusNotFound, map[string]string{"detail": "Translation not found"})
		return
	}

	WriteJSON(w, http.StatusOK, translationStatusResponse{
		TranslationID: item.ID,
		Status:        item.Status,
		Progress:      intPtrIfKnown(item.Progress, item.Status),
		Total:         intPtrIfKnown(item.Total, item.Status),
	})
}

func DeleteTranslation(w http.ResponseWriter, r *http.Request) {
	if err := validateDependencies(); err != nil {
		WriteJSON(w, http.StatusInternalServerError, map[string]string{"detail": err.Error()})
		return
	}

	translationID := pathParam(r, "translation_id")
	if !sharedStore.Delete(translationID) {
		WriteJSON(w, http.StatusNotFound, map[string]string{"detail": "Translation not found"})
		return
	}
	sharedQueue.CleanupProgress(translationID)

	WriteJSON(w, http.StatusOK, map[string]bool{"ok": true})
}

func TranslationStream(w http.ResponseWriter, r *http.Request) {
	if err := validateDependencies(); err != nil {
		WriteJSON(w, http.StatusInternalServerError, map[string]string{"detail": err.Error()})
		return
	}

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.WriteHeader(http.StatusOK)

	flusher, ok := w.(http.Flusher)
	if !ok {
		emitSSE(w, map[string]any{"type": "error", "message": "Streaming is not supported"})
		return
	}

	translationID := pathParam(r, "translation_id")
	item, exists := sharedStore.Get(translationID)
	if !exists {
		emitSSE(w, map[string]any{"type": "error", "message": "Translation not found"})
		flusher.Flush()
		return
	}

	if item.Status == "failed" {
		emitSSE(w, map[string]any{"type": "error", "message": derefOr(item.ErrorMessage, "Translation failed")})
		flusher.Flush()
		return
	}

	if item.Status == "completed" {
		replayCompletedStream(w, flusher, item)
		sharedQueue.CleanupProgress(translationID)
		return
	}

	sharedQueue.StartProcessing(translationID)
	streamLiveProgress(r.Context(), w, flusher, translationID)
}

func streamLiveProgress(ctx context.Context, w http.ResponseWriter, flusher http.Flusher, translationID string) {
	ticker := time.NewTicker(20 * time.Millisecond)
	defer ticker.Stop()

	startSent := false
	lastProgress := 0

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			item, exists := sharedStore.Get(translationID)
			if !exists {
				emitSSE(w, map[string]any{"type": "error", "message": "Translation not found"})
				flusher.Flush()
				return
			}

			if item.Status == "failed" {
				emitSSE(w, map[string]any{"type": "error", "message": derefOr(item.ErrorMessage, "Translation failed")})
				flusher.Flush()
				sharedQueue.CleanupProgress(translationID)
				return
			}

			progress, ok := sharedQueue.GetProgress(translationID)
			if !ok {
				continue
			}

			if !startSent && progress.Total > 0 {
				emitSSE(w, map[string]any{
					"type":           "start",
					"translation_id": translationID,
					"total":          progress.Total,
					"paragraphs":     paragraphInfo(item.Paragraphs),
				})
				flusher.Flush()
				startSent = true
			}

			for i := lastProgress; i < len(progress.Results); i++ {
				result := progress.Results[i]
				emitSSE(w, map[string]any{
					"type":    "progress",
					"current": i + 1,
					"total":   progress.Total,
					"result": map[string]any{
						"segment":         result.Segment,
						"pinyin":          result.Pinyin,
						"english":         result.English,
						"index":           result.Index,
						"paragraph_index": result.ParagraphIndex,
					},
				})
				flusher.Flush()
			}
			lastProgress = len(progress.Results)

			if progress.Status == "completed" || item.Status == "completed" {
				fresh, _ := sharedStore.Get(translationID)
				emitSSE(w, map[string]any{
					"type":            "complete",
					"paragraphs":      fresh.Paragraphs,
					"fullTranslation": fresh.FullTranslation,
				})
				flusher.Flush()
				sharedQueue.CleanupProgress(translationID)
				return
			}
		}
	}
}

func replayCompletedStream(w http.ResponseWriter, flusher http.Flusher, item translation.Translation) {
	emitSSE(w, map[string]any{
		"type":            "start",
		"translation_id":  item.ID,
		"total":           item.Total,
		"paragraphs":      paragraphInfo(item.Paragraphs),
		"fullTranslation": item.FullTranslation,
	})
	flusher.Flush()

	current := 0
	for paraIdx, para := range item.Paragraphs {
		for _, seg := range para.Translations {
			current++
			emitSSE(w, map[string]any{
				"type":    "progress",
				"current": current,
				"total":   item.Total,
				"result": map[string]any{
					"segment":         seg.Segment,
					"pinyin":          seg.Pinyin,
					"english":         seg.English,
					"index":           current - 1,
					"paragraph_index": paraIdx,
				},
			})
			flusher.Flush()
		}
	}

	emitSSE(w, map[string]any{
		"type":            "complete",
		"paragraphs":      item.Paragraphs,
		"fullTranslation": item.FullTranslation,
	})
	flusher.Flush()
}

func emitSSE(w http.ResponseWriter, payload any) {
	data, err := json.Marshal(payload)
	if err != nil {
		_, _ = fmt.Fprint(w, "data: {\"type\":\"error\",\"message\":\"Failed to encode SSE payload\"}\n\n")
		return
	}
	_, _ = fmt.Fprintf(w, "data: %s\n\n", data)
}

func paragraphInfo(paragraphs []translation.ParagraphResult) []map[string]any {
	out := make([]map[string]any, 0, len(paragraphs))
	for _, para := range paragraphs {
		out = append(out, map[string]any{
			"segment_count": len(para.Translations),
			"indent":        para.Indent,
			"separator":     para.Separator,
		})
	}
	return out
}

func intPtrIfKnown(value int, status string) *int {
	if status == "pending" && value == 0 {
		return nil
	}
	v := value
	return &v
}

func previewPtr(value *string, max int) *string {
	if value == nil {
		return nil
	}
	out := preview(*value, max)
	return &out
}

func derefOr(v *string, fallback string) string {
	if v == nil || *v == "" {
		return fallback
	}
	return *v
}
